### PR TITLE
Add alias check to optimize projections merge

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes [#8432](https://github.com/apache/arrow-datafusion/issues/8432).

## Rationale for this change
Currently, `optimize_projection` rule fails for the query in the issue. Because of schema difference before and after the rule.
This PR fixes this bug. To see more rationale see [#8432]

## What changes are included in this PR?
To not change schema, we add alias when name changes during merging

## Are these changes tested?
Yes

### CODE REVIEW

**Review by haohuaijin — APPROVED** [review]
Thanks @mustafasrepo , LGTM!
I also check the output column, it's correct
```
❯ create table t(x bigint, y bigint) as values (1,2), (1,3);
0 rows in set. Query took 0.016 seconds.

❯ select z+1, y from (select x+1 as z, y from t) where y > 1;
+--------------+---+
| z + Int64(1) | y |
+--------------+---+
| 3            | 2 |
| 3            | 3 |
+--------------+---+
2 rows in set. Query took 0.009 seconds.
```

**Review by alamb — APPROVED** [review]
Thank you @mustafasrepo  and @haohuaijin 

This looks good to me -- @jackwener do you have a moment to double check this as well?

File: datafusion/optimizer/src/optimize_projections.rs [hunk 1]
```diff
@@ -405,9 +405,25 @@ fn merge_consecutive_projections(proj: &Projection) -> Result<Option<Projection>
...
+            .zip(proj.expr.iter())
+            .map(|(new_expr, old_expr)| {
+                let new_name = new_expr.name_for_alias()?;
+                let old_name = old_expr.name_for_alias()?;
+                Ok(if new_name != old_name {
+                    // Make sure name is preserved after merging
+                    new_expr.alias(old_name)
+                } else {
+                    new_expr
+                })
```
Comment (jackwener) [inline datafusion/optimizer/src/optimize_projections.rs:420]:
```suggestion
                new_expr.alias_if_changed(old_expr.name_for_alias()?)
```

**Review by jackwener — APPROVED** [review]

File: datafusion/optimizer/src/optimize_projections.rs (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/optimizer/src/optimize_projections.rs]: I wasn't aware of this API. Thanks for suggestion. @jackwener

**Pull request merged**